src/util/result.h

```diff
@@ -8,16 +8,133 @@
 #include <attributes.h>
 #include <util/translation.h>
 
+#include <memory>
+#include <optional>
+#include <tuple>
 #include <variant>
+#include <vector>
 
 namespace util {
+namespace detail {
+//! Empty string list
+const std::vector<bilingual_str> EMPTY_LIST{};
 
+//! Helper to move elements from one container to another.
+template <typename T>
+void MoveElements(T& src, T& dest)
+{
+    dest.insert(dest.end(), std::make_move_iterator(src.begin()), std::make_move_iterator(src.end()));
+    src.clear();
+}
+
+//! Error information only allocated if there are errors or warnings.
+template <typename F>
+struct ErrorInfo {
+    std::optional<std::conditional_t<std::is_same<F, void>::value, std::monostate, F>> failure;
+    std::vector<bilingual_str> errors;
+    std::vector<bilingual_str> warnings;
+};
+
+//! Result base class which is inherited by Result<T, F>.
+template <typename T, typename F>
+class ResultBase;
+
+//! Result base specialization for empty (T=void) value type. Holds error
+//! information and provides accessor methods.
+template <typename F>
+class ResultBase<void, F>
+{
+protected:
+    std::unique_ptr<ErrorInfo<F>> m_info;
+
+    template <typename... Args>
+    void InitFailure(Args&&... args)
+    {
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->failure.emplace(std::forward<Args>(args)...);
+    }
+    void InitValue() {}
+    template <typename O>
+    void MoveValue(O& other) {}
+    void DestroyValue() {}
+
+public:
+    void AddError(bilingual_str error)
+    {
+        if (error.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->errors.emplace_back(std::move(error));
+    }
+
+    void AddWarning(bilingual_str warning)
+    {
+        if (warning.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->warnings.emplace_back(std::move(warning));
+    }
+
+    //! Success check.
+    operator bool() const { return !m_info || !m_info->failure; }
+
+    //! Error retrieval.
+    template <typename _F = F>
+    std::enable_if_t<!std::is_same<_F, void>::value, const _F&> GetFailure() const { assert(!*this); return *m_info->failure; }
+    const std::vector<bilingual_str>& GetErrors() const { return m_info ? m_info->errors : EMPTY_LIST; }
+    const std::vector<bilingual_str>& GetWarnings() const { return m_info ? m_info->warnings : EMPTY_LIST; }
+};
+
+//! Result base class for T value type. Holds value and provides accessor methods.
+template <typename T, typename F>
+class ResultBase : public ResultBase<void, F>
+{
+protected:
+    union { T m_value; };
```

Comment:
I don't understand why the `m_value` is in a `union` while there is only one member in the `union`?

Reply:
> I don't understand why the `m_value` is in a `union` while there is only one member in the `union`? Would you explain why?

I'll add a code comment. But using a union avoids `m_value` getting constructor and destructor being called automatically, so in the failure case `m_value` is never constructed.